src/tests/regress/data/sql/strings.sql

```diff
@@ -137,131 +137,136 @@ SELECT SUBSTRING('string' FROM -10 FOR 2147483646) AS "string";
+--@ SELECT SUBSTRING('abcdefg' SIMILAR 'a#"%|ab#"g' ESCAPE '#') AS "bcdef";
+--@ 
+--@ -- Can't have more than two part separators
+--@ SELECT SUBSTRING('abcdefg' SIMILAR 'a*#"%#"g*#"x' ESCAPE '#') AS "error";
+--@ 
+--@ -- Postgres extension: with 0 or 1 separator, assume parts 1 and 3 are empty
+--@ SELECT SUBSTRING('abcdefg' SIMILAR 'a#"%g' ESCAPE '#') AS "bcdefg";
+--@ SELECT SUBSTRING('abcdefg' SIMILAR 'a%g' ESCAPE '#') AS "abcdefg";
+--@ 
+--@ -- substring() with just two arguments is not allowed by SQL spec;
+--@ -- we accept it, but we interpret the pattern as a POSIX regexp not SQL
+--@ SELECT SUBSTRING('abcdefg' FROM 'c.e') AS "cde";
+--@ 
+--@ -- With a parenthesized subexpression, return only what matches the subexpr
+--@ SELECT SUBSTRING('abcdefg' FROM 'b(.*)f') AS "cde";
 
 -- Check behavior of SIMILAR TO, which uses largely the same regexp variant
-SELECT 'abcdefg' SIMILAR TO '_bcd%' AS true;
-SELECT 'abcdefg' SIMILAR TO 'bcd%' AS false;
-SELECT 'abcdefg' SIMILAR TO '_bcd#%' ESCAPE '#' AS false;
-SELECT 'abcd%' SIMILAR TO '_bcd#%' ESCAPE '#' AS true;
--- Postgres uses '\' as the default escape character, which is not per spec
-SELECT 'abcdefg' SIMILAR TO '_bcd\%' AS false;
--- and an empty string to mean "no escape", which is also not per spec
-SELECT 'abcd\efg' SIMILAR TO '_bcd\%' ESCAPE '' AS true;
--- these behaviors are per spec, though:
-SELECT 'abcdefg' SIMILAR TO '_bcd%' ESCAPE NULL AS null;
-SELECT 'abcdefg' SIMILAR TO '_bcd#%' ESCAPE '##' AS error;
-
--- Test back reference in regexp_replace
-SELECT regexp_replace('1112223333', E'(\\d{3})(\\d{3})(\\d{4})', E'(\\1) \\2-\\3');
-SELECT regexp_replace('AAA   BBB   CCC   ', E'\\s+', ' ', 'g');
-SELECT regexp_replace('AAA', '^|$', 'Z', 'g');
-SELECT regexp_replace('AAA aaa', 'A+', 'Z', 'gi');
--- invalid regexp option
-SELECT regexp_replace('AAA aaa', 'A+', 'Z', 'z');
+--@ SELECT 'abcdefg' SIMILAR TO '_bcd%' AS true;
+--@ SELECT 'abcdefg' SIMILAR TO 'bcd%' AS false;
+--@ SELECT 'abcdefg' SIMILAR TO '_bcd#%' ESCAPE '#' AS false;
+--@ SELECT 'abcd%' SIMILAR TO '_bcd#%' ESCAPE '#' AS true;
+--@ -- Postgres uses '\' as the default escape character, which is not per spec
+--@ SELECT 'abcdefg' SIMILAR TO '_bcd\%' AS false;
+--@ -- and an empty string to mean "no escape", which is also not per spec
+--@ SELECT 'abcd\efg' SIMILAR TO '_bcd\%' ESCAPE '' AS true;
+--@ -- these behaviors are per spec, though:
+--@ SELECT 'abcdefg' SIMILAR TO '_bcd%' ESCAPE NULL AS null;
+--@ SELECT 'abcdefg' SIMILAR TO '_bcd#%' ESCAPE '##' AS error;
+
+-- Test backslash escapes in regexp_replace's replacement string
+--@ SELECT regexp_replace('1112223333', E'(\\d{3})(\\d{3})(\\d{4})', E'(\\1) \\2-\\3');
+--@ SELECT regexp_replace('foobarrbazz', E'(.)\\1', E'X\\&Y', 'g');
+--@ SELECT regexp_replace('foobarrbazz', E'(.)\\1', E'X\\\\Y', 'g');
+--@ -- not an error, though perhaps it should be:
+--@ SELECT regexp_replace('foobarrbazz', E'(.)\\1', E'X\\Y\\1Z\\');
+--@ 
```

Comment:
The `expected` is from PostgreSQL 15 but `sql` was from PostgreSQL 14. So we add the extra tests from PostgreSQL 15 to `sql` as well.